pm/parser/perl_extensions/man/man3/PIF_2_LIF.3
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH PIF_2_LIF 3 "perl 5.004" "30/Nov/100" "User Contributed Perl Documentation"
.IX Title "PIF_2_LIF 3"
.UC
.IX Name "PIF_2_LIF - A Perl extension that converts files it PIF/LIF  without doing anything special 
              to records in a Parser Intermediate file."
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
PIF_2_LIF \- A Perl extension that converts files it PIF/LIF  without doing anything special 
              to records in a Parser Intermediate file.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.PP
.Vb 1
\&  use PIF_2_LIF;
.Ve
.Vb 1
\&  $obj_id = PIF_2_LIF->New();
.Ve
.Vb 1
\&  $obj_id->load_config( $keep_files, $debug, \e%config )
.Ve
.Vb 1
\&  $obj_id->process( $in_dir, \e@in_files, $out_dir )
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.PP
.Vb 6
\&  There are three entry points to this object.  The first 'New' creates the
\&  ibject and stores the identifier in the scalar varaible '$objid'.  The second
\&  'load_config' allows the user to configure the object to run as they desire.
\&  The third 'process' goes through and processes the files in the input directory
\&  '$in_dir' as defined by the configuration information.  The output
\&  of the process is output into the directory '$out_dir'.
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
D. Fisher, david.fisher@adc.metrica.co.uk
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\fIperl\fR\|(1).

.rn }` ''


pm/parser/perl_extensions/lib/site_perl/PIF_2_LIF.pm
#
#-------------------------------------------------------------------------------
# @(#) common-modules/PIF_2_LIF/PIF_2_LIF.pm common-modules_r2.1.3.2:cset.000563:4:4 06/07/99 @(#)
#-------------------------------------------------------------------------------
#
#
package PIF_2_LIF;
use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Exporter;
require AutoLoader;

@ISA = qw(Exporter AutoLoader);
# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.
@EXPORT = qw(
	
);
$VERSION = '0.01';

#use diagnostics;
use GenUtils;
use File::Basename;
use LIF_Writer;
use PIF_Handler;
use AudLog;

sub New {
	my $proto = shift;
	my $class = ref($proto) || $proto;
	my $self = {};

	# These two variables should be set to true of false
	# depending on whether or not the desired type of output file
	# is wanted
	$self->{PRODUCE_PIF} = 0;
	$self->{PRODUCE_LIF} = "True";
	$self->{keep_files} = "True";
	$self->{debug} = "True";

	$self->{REDUNDANT_COUNTERS} = ();

	$self->{NEW_HEADER_NAMES} = ();
	$self->{NEW_HEADER_VALUES} = ();

	bless ($self, $class);
}

sub load_config {
	my $self = shift;
	($self->{keep_files}, $self->{debug}, $self->{'__config__'}) = @_;

	# Inserting all the configuration information into the objects records
	my $key;
	foreach $key ( keys %{$self->{'__config__'}} ) {
		$self->{$key} = $self->{'__config__'}->{$key};
	}
}

# Object Methods
sub process {
	my ($self, $in_dir, $in_files, $out_dir) = @_;

	# Declaring variables local to this subroutine
	my ($num_succ_processed, $dir, @files );
	my ($pattern, %files_h);

	$num_succ_processed = 0;

	# Get the list of files that are to be processed by this rule.  But first
	# checking if they have provided a list of regular expressions or just
	# one.
	if( ref($self->{INPUT_FILE_DESCRIPTION}) eq "ARRAY" ) {
		foreach $pattern ( @{$self->{INPUT_FILE_DESCRIPTION}} ) {
			push @files, grep /$pattern/, @$in_files;
		}
	}
	else {
		@files = grep /$self->{INPUT_FILE_DESCRIPTION}/, @$in_files;
	}

	# Removing duplicates from the list of files
	@files_h{@files} = ();

	# If the OUTPUT_DIR is specified set the output directory
	# to be OUTPUT_DIR, rather than the one specified on the 
	# command line.  first check that it exists and is writable etc.
	# Otherwise, log and error message and exit this rule.
	if( exists($self->{OUTPUT_DIR}) && $self->{OUTPUT_DIR} ) {
	    $dir = $self->{OUTPUT_DIR};
		if( -d $dir && -w $dir && -r $dir && -x $dir ) {
			LogMess("For this rule, output directory set to '".$dir."'", 3);
			$out_dir = $self->{OUTPUT_DIR};
		}
		else {
			LogMess("ERROR: Output directory '".$dir."' has incorrect permissions", 1);
			return -1;
		}
	}

	# Loop over the list of files to process for this rule
	foreach ( keys %files_h ) {
		if( $self->process_a_file($in_dir."/".$_, $out_dir) != 0 ) {
			AudMess("ERROR: processing '$_'");
		}
		$num_succ_processed++;
	}

	return $num_succ_processed;
}


sub process_a_file() {
	my ($self, $in_file, $out_dir) = @_;

	my ($i_obj, $o_obj, $blk_name, $out_file, $new_pif, $num_names, $num_vals);
	my (@names, @values);
	my (%h_i, %counter_hash);

	AudMess("Processing '${in_file}'");

	$i_obj = PIF_Handler->New();
	$i_obj->Open($in_file);

	# Reading the Header information.
	$blk_name = $i_obj->Read_Names(\@names);
	$i_obj->Read_Values(\@values);

	@h_i{@names} = @values;
	if (( $self->{NEW_HEADER_NAMES})||($self->{NEW_HEADER_VALUES})){
		# count number of names & values (assign to $number
		# if uneven match, print error and abort addition

		$num_names=@{$self->{NEW_HEADER_NAMES}};
		$num_vals=@{$self->{NEW_HEADER_VALUES}};

		if ($num_names == $num_vals){
			LogMess("  Added $num_names new name(s) and value(s) to HEADER record", 5);
			@h_i{@{$self->{NEW_HEADER_NAMES}}}=@{$self->{NEW_HEADER_VALUES}};
		}
		else{
			LogMess("  Mismatched No. of NAME($num_names) and VALUE($num_vals) to add to header", 2); 
			return -1;
		}
	}

	# Looping over all the data blocks in the file
	while( $blk_name = $i_obj->Read_Names(\@names) ) {

		# Creating a name for output block name if not provided
		if( exists($self->{OUTPUT_BLOCK_NAME}) &&
			$self->{OUTPUT_BLOCK_NAME}) {
			$blk_name = $self->{OUTPUT_BLOCK_NAME};
		}

		if (not $o_obj){

			# Creating the output filename
			$out_file = basename($in_file);
			$out_file =~ s/\.pif$/-#-P.pt/;

			# Appending an optional user specified string to the
			# start of the output file name.
			$out_file = $self->{OUTPUT_FILENAME_START}.'-#-'.$out_file
				if( exists($self->{OUTPUT_FILENAME_START})
				&& $self->{OUTPUT_FILENAME_START});

			$new_pif = dirname($in_file)."/".$out_file;

			if( $self->{PRODUCE_PIF} ) {
				$i_obj->Open_PIF_Write($new_pif);
				$i_obj->PIF_Write("HEADER", \%h_i);
			}

			AudMess("Output file: '${out_file}'");

			if( $self->{PRODUCE_LIF} ) {
				$o_obj = LIF_Writer->New();
				$o_obj->Open($out_dir."/".$out_file);

				$o_obj->Open_Block();
				$o_obj->Block_Info(\%h_i);
				$o_obj->WriteToFile();
			}
		}
	          
		# Initialising the hash before inserting a different
		# set of names from this block.
		%counter_hash = ();

		# Looping over all the Data lines in this block
		while ( $i_obj->Read_Values(\@values) ) {

			@counter_hash{@names}=@values;

			# Delete the counters from the hash that now make no sense
			foreach ( @{$self->{REDUNDANT_COUNTERS}} ) {
				delete $counter_hash{$_};
			}

			# Writing the Block into the LIF file
			$o_obj->Create_Block($blk_name, \%counter_hash) if $self->{PRODUCE_LIF};

			# Writing the block into the new PIF file
			$i_obj->PIF_Write($blk_name, \%counter_hash) if $self->{PRODUCE_PIF};
		}
	}

	# Closing the Output file and renaming the completed output file to
	# end in the appropriate extension
	
	if($self->{PRODUCE_PIF}&& $i_obj )   {
		$i_obj->Close_PIF_Write();
		rename_completed_file(".pt", ".pif", $new_pif);
	}
	if($self->{PRODUCE_LIF}&& $o_obj ) {
		$o_obj->Close();
		# add read privilege to metrica on sdwg1
		# add by fanjun on 2003-11-07
		chmod(0644,$out_dir."/".$out_file);
		#AudMess("Modify Privilege of $out_dir/$out_file to 0644");
		store_processed_file($self->{keep_files}, 0, $self->{debug}, rename_completed_file(".pt", ".lif", $out_dir."/".$out_file));
	}

	return 0;
}

1; # So the use or require succeeds
__END__
# Below is the stub of documentation for your module. You better edit it!

=head1 NAME

PIF_2_LIF - A Perl extension that converts files it PIF/LIF  without doing anything special 
              to records in a Parser Intermediate file.

=head1 SYNOPSIS

  use PIF_2_LIF;

  $obj_id = PIF_2_LIF->New();

  $obj_id->load_config( $keep_files, $debug, \%config )

  $obj_id->process( $in_dir, \@in_files, $out_dir )

=head1 DESCRIPTION

  There are three entry points to this object.  The first 'New' creates the
  ibject and stores the identifier in the scalar varaible '$objid'.  The second
  'load_config' allows the user to configure the object to run as they desire.
  The third 'process' goes through and processes the files in the input directory
  '$in_dir' as defined by the configuration information.  The output
  of the process is output into the directory '$out_dir'.

=head1 AUTHOR

D. Fisher, david.fisher@adc.metrica.co.uk

=head1 SEE ALSO

perl(1).

=cut
